third_party/move/move-compiler-v2/src/experiments.rs

```diff
@@ -2,42 +2,186 @@
 // Parts of the project are originally copyright © Meta Platforms, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-/// Container for experiments in the compiler. Those can be activated
-/// via the `--experiment=<name>` flag. One can also use the env var
-/// `MVC_EXP=<comma list>` to activate those flags.
-///
-/// Moreover, one can activate an experiment in a test source by using a
-/// comment as such in a unit test:
-/// ```text
-///   // experiment: <name>
-/// ```
-/// This can be repeated an arbitrary number of times. The test will then be run for
-/// the default configuration and for each of the named experiments separately (if it is a
-/// baseline test, a different baseline file will be generated each time).
-///
-/// Each new experiment should have a description and explicit note about its retention.
-///
-/// - Permanent: the experiment is available indefinitely
-/// - Temporary: the experiment is intended to be removed after some time. Please document
-///   the condition under which it can be removed.
-pub struct Experiment();
+//! Defines 'experiments' (flags) for the compiler. Most phases of the
+//! compiler can be enabled or disabled via an experiment. An experiment
+//! can be set via the command line (`--experiment name[=on/off]`),
+//! via an environment variable (`MVC_EXP=def,..` where `def` is
+//! `name[=on/off]`), or programmatically. Experiments are retrieved
+//! via `options.experiment_on(NAME)`.
+//!
+//! The declaration of experiments happens via the datatype `Experiment`
+//! which defines its name, description, and default value. The default
+//! can be either be a fixed constant or inherited from another
+//! experiment, effectively allowing to activate a group of experiments
+//! via some meta-experiment. For example, the `OPTIMIZE` experiment
+//! turns on or off a bunch of other experiments, unless those are
+//! defined explicitly.
 
+use once_cell::sync::Lazy;
+use std::collections::BTreeMap;
+
+#[derive(Clone)]
+pub struct Experiment {
+    /// The name of the experiment
+    pub name: String,
+    /// A description of the experiment
+    pub description: String,
+    /// Whether the default is true or false
+    pub default: DefaultValue,
+}
+
+#[derive(Clone)]
+pub enum DefaultValue {
+    /// Whether the default is a fixed value.
+    Given(bool),
+    /// Whether the default is inherited from another experiment
+    Inherited(String),
+}
+
+pub static EXPERIMENTS: Lazy<BTreeMap<String, Experiment>> = Lazy::new(|| {
+    use DefaultValue::*;
+    let experiments = vec![
+        Experiment {
+            name: Experiment::CHECKS.to_string(),
+            description: "Turns on or off a group of context checks".to_string(),
```

Comment:
Does an "Experiment" represent a single experimental case vs the default?

If so, the description should say "Turns **off** a group of context checks", as that is the experiment condition.  Saying "on or off" leaves me confused about the experimental condition.  Same for the below.

Reply:
We also may want to have the names reflect what's happening in the experiment condition, e.g., "NO_CHECKS".

Reply:
Oh, these are now "flags", not really experiments.  names matter.

Reply:
Experiments have values as well, so this wouldn't be the reason why to change the name. In fact at Google various experiment frameworks have even more then boolean, there can be enums etc. (Look at the experiments of Chrome.)

If you'd read the PR description this confusing could have been avoided, that they are on/off now is prominently documented there, also in `experiments.rs` ;-)